infix/lexer.go

```diff
@@ -0,0 +1,225 @@
+	return l.tokens
+}
+
+// run the state machine
+func (l *lexer) run() {
+	for state := lexStart; state != nil; {
+		state = state(l)
+	}
+	close(l.tokens)
+}
+
+// emit a token.
+func (l *lexer) emit(tok Token) {
+	l.tokens <- Tokval{
+		T: tok,
+		V: l.input[l.start:l.pos],
+	}
+	l.start = l.pos
+}
+
+// errorf emits an illegal token. This token carries
+// the lexer error.
+func (l *lexer) errorf(msg string, args ...interface{}) stateFn {
+	l.tokens <- Tokval{
+		T: Illegal,
+		V: fmt.Sprintf(msg, args...),
+	}
+	return nil
+}
+
+// peek looks for the next rune from the input
+// but do not increases the cursor.
+// It returns eof when reaches the end of input.
+func (l *lexer) peek() rune {
+	r := l.next()
+	if r == eof {
+		return eof
+	}
+	l.backup()
+	return r
+}
+
+// next returns the next rune from input or
+// eof if reached end of input.
+func (l *lexer) next() (r rune) {
+	if l.pos >= len(l.input) {
+		// zeroes the width to simplify handling at
+		// end of input. Some functions like acceptRun*
+		// do a explicit rewind/backup of last processed
+		// character, but in case of eof we dont want
+		// that.
+		l.width = 0
+		return eof
+	}
+	r, l.width = utf8.DecodeRuneInString(l.input[l.pos:])
+	l.pos += l.width
+	return r
+}
+
+// backup rollback one rune.
+func (l *lexer) backup() {
+	l.pos -= l.width
+}
+
+// ignore bytes processed since last token.
+func (l *lexer) ignore() {
+	l.start = l.pos
+}
+
+// accept checks if next is in the valid string.
+func (l *lexer) accept(valid string) bool {
+	return strings.IndexRune(valid, l.peek()) != -1
+}
+
+// acceptRun consumes next runes if valid.
+func (l *lexer) acceptRun(valid string) {
+	for strings.IndexRune(valid, l.next()) >= 0 {
+	}
+
+	l.backup()
+}
+
+// acceptRunfn is like acceptRun but uses fn as 
+// function comparator.
+func (l *lexer) acceptRunfn(fn func(r rune) bool) {
+	for fn(l.next()) {
+
+	}
+
+	l.backup()
+}
+
+// lexStart is the start of state machine
+func lexStart(l *lexer) stateFn {
+	r := l.next()
+	switch {
+	case unicode.IsSpace(r):
+		l.acceptRunfn(unicode.IsSpace)
+		l.ignore()
+		return lexStart
+	case r == eof: 
+		return nil
+	case r >= '0' && r <= '9':
+		l.backup()
+		return lexNumber
+	case r == '|':
+		l.emit(OR)
+		return lexStart
+	case r == '&':
+		l.emit(AND)
+		return lexStart
+	case r == '^':
+		l.emit(XOR)
+		return lexStart
+	case r == '~':
+		l.emit(NOT)
+		return lexStart
+	case r == '<':
+		next := l.next()
+		if next != '<' {
+			return l.errorf("unexpected %q", next)
+		}
+		l.emit(SHL)
+		return lexStart
+	case r == '>':
+		next := l.next()
+		if next != '>' {
+			return l.errorf("unexpected %q", next)
+		}
+		l.emit(SHR)
+		return lexStart
+	case r == '(':
+		l.emit(LParen)
+		return lexStart
+	case r == ')':
+		l.emit(RParen)
+		return lexStart
+	default:
+		return l.errorf("Unexpected %q at %d", r, l.pos)
+	}
+	return nil
+}
+
+// lexer of dec, hex and bin numbers.
+func lexNumber(l *lexer) stateFn {
+	r := l.next()
+	next := l.peek()
```

Comment:
This function is called 'lexNumber' and binary/hex is only handled here.
If there existed other places where handleBinary or handleHex was needed, I'm sure I would have put it on separate functions.

Reply:
I like to use functions to avoid having to write comments. that seems to be the case here (not just because the function is going to be used on more than one place). Having to comment what a function is doing inside it is usually a hint that I could extract a function to say that, specially if the overhead is negligible.